### PR TITLE
feat: metadata will render recipe with context

### PR DESCRIPTION
**Pull request by nichmor**
This PR aims to give `Metadata` ability to render recipe using context values.

It will be used by conda-smithy to ask metadata.name() and metadata.version()

### CODE REVIEW

File: src/rattler_build_conda_compat/render.py [hunk 1]
```diff
@@ -59,10 +64,53 @@ def __init__(
...
-        with open(os.path.join(self.path, self._meta_name), "r") as recipe_yaml:
-            return yaml.load(recipe_yaml)
+    def parse_recipe(self) -> dict[str, Any]:
+        recipe_path: Path = Path(self.path) / self._meta_name
+
+        yaml_content = load_yaml(recipe_path.read_text())
+
+        return render_recipe_with_context(yaml_content)
+
+    def name(self) -> str:
```
Comment (nichmor) [inline src/rattler_build_conda_compat/render.py:74]: this override name() and version() from conda_build.Metadata base class. typing override was introduced in 3.12 so I cannot use it here

**Review by baszalmstra — COMMENTED** [review]
Does this reflect the same API that `CondaMetaData` has? Or would you access the `name` and `version` differently?

File: src/rattler_build_conda_compat/render.py [hunk 2]
```diff
@@ -59,10 +64,53 @@ def __init__(
...
+        Otherwise it will return the name of the `package` field.
+
+        Raises:
+            - CondaBuildUserError: If the `name` contains bad characters.
+
+        """
+        name = _get_recipe_metadata(self.meta, "name")
+
+        if not name:
+            sys.exit(f"Error: package/name missing in: {self.meta_path!r}")
```
[the remark this answers is not in the document]
Comment (nichmor) [reply src/rattler_build_conda_compat/render.py:88]: This wasn't my implementation - I tried to reuse the same approach as conda_build does - so I decided to not move from sys.exit. Let me know if you think that it's worth to change it

File: src/rattler_build_conda_compat/render.py [hunk 3]
```diff
@@ -59,10 +64,53 @@ def __init__(
...
+        Otherwise it will return the version of the `package` field.
+
+        Raises:
+            - CondaBuildUserError: If the `version` contains bad characters.
+            - ValueError: If the version starts with a period.
+        """
+        version = _get_recipe_metadata(self.meta, "version")
+
+        if not version:
+            sys.exit(f"Error: package/version missing in: {self.meta_path!r}")
```
[the remark this answers is not in the document]
Comment (nichmor) [reply src/rattler_build_conda_compat/render.py:110]: ^ same here

File: src/rattler_build_conda_compat/render.py [hunk 4]
```diff
@@ -59,10 +64,53 @@ def __init__(
...
-    def parse_recipe(self):
-        yaml = YAML()
-        with open(os.path.join(self.path, self._meta_name), "r") as recipe_yaml:
-            return yaml.load(recipe_yaml)
+    def parse_recipe(self) -> dict[str, Any]:
+        recipe_path: Path = Path(self.path) / self._meta_name
+
+        yaml_content = load_yaml(recipe_path.read_text())
+
+        return render_recipe_with_context(yaml_content)
```
[the remark this answers is not in the document]
Comment (nichmor) [reply src/rattler_build_conda_compat/render.py:72]: yes, it is used here; https://github.com/prefix-dev/rattler-build-conda-compat/pull/41/files#diff-77478d772015c6eb0acf745f154a3170ff44a128dca1e00997c90739d41c54dbR55

**Comment by nichmor** [comment]
> Does this reflect the same API that `CondaMetaData` has? Or would you access the `name` and `version` differently?

it's the same API - I just override it because we extract them from recipe/package depending on `outputs` field

File: src/rattler_build_conda_compat/render.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nichmor) [reply src/rattler_build_conda_compat/render.py]: changed.

File: src/rattler_build_conda_compat/render.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nichmor) [reply src/rattler_build_conda_compat/render.py]: changed

File: src/rattler_build_conda_compat/render.py [hunk 5]
```diff
@@ -59,10 +64,56 @@ def __init__(
...
+            - ValueError: If the version starts with a period or version is missing.
+        """
+        version = _get_recipe_metadata(self.meta, "version")
+
+        if not version:
+            raise ValueError(f"Error: package/version missing in: {self.meta_path!r}")
+
+        check_bad_chrs(version, "package/version")
+        if version.startswith("."):
+            raise ValueError("Fully-rendered version can't start with period -  got %s", version)
```
Comment (baszalmstra) [inline src/rattler_build_conda_compat/render.py:115]: Maybe also use fstrings here like you did for the other exceptions.

**Review by baszalmstra — APPROVED** [review]

**Pull request merged**